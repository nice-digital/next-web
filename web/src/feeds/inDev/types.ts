import {
	AreaOfInterest,
	ProductGroup,
	ProductTypeAcronym,
} from "@/feeds/publications/types";

export enum FeedPath {
	AllProjects = "/gidprojects/all",
	InConsultationProjects = "/inconsultationprojects",
	ProjectDetail = "/project/",
}

export enum TopicSelectionReason {
	"Monitor" = "Monitor - awaiting publication of further literature or evidence",
	"Anticipate" = "Anticipate the topic will be of importance to patients, carers, professionals, commissioners and the health of the public to ensure clinical benefit is realised, inequalities in use addressed, and help them make the best use of NHS resources",
	"NotEligible" = "Not eligible for Health Technology Evaluation guidance",
	"FurtherDiscussion" = "Further discussion required at the Topic Selection Oversight Panel (TSOP)",
}

export enum ProjectStatus {
	"Complete" = "Complete",
	"Discontinued" = "Discontinued",
	"Suspended" = "Suspended",
	"Referred" = "Referred",
	"InProgress" = "InProgress",
	"Proposed" = "Proposed",
	"TopicSelection" = "TopicSelection",
	"ImpactedByCOVID19" = "ImpactedByCOVID19",
}

/** Not a definitive list - this may change over time */
export enum ProjectProcess {
	"MTA" = "MTA",
	"IP" = "IP",
	"TAG" = "TAG",
	"MT" = "MT",
	"DT" = "DT",
	"STAPre2018" = "STA pre-2018",
	"STA2018" = "STA 2018",
	"QSD" = "QSD",
	"CG" = "CG",
	"APG" = "APG",
	"SG" = "SG",
	"HST" = "HST",
	"MIB" = "MIB",
	"STAReview" = "STA Review",
	"PHG" = "PHG",
	"MTAReview" = "MTA Review",
	"CDFReview" = "CDF Review",
	"NG" = "NG",
	"APA" = "APA",
	"ES" = "ES",
	"SC" = "SC",
	"COVID19RapidGuideline" = "COVID-19 rapid guideline",
}

export enum ProjectGroup {
	Guideline = ProductGroup.Guideline,
	Guidance = ProductGroup.Guidance,
	Advice = ProductGroup.Advice,
	Standard = ProductGroup.Standard,
	Other = ProductGroup.Other,
}

export type GIDReference = `GID-${Uppercase<string>}${number}`;

export interface AllProjects {
	links: {
		// In theory InDev supports pagination, but in reality the gidprojects/all feed always returns all projects and a page size of 2048
		first: [
			{
				href: "/gidprojects/All?Page=0&PageSize=2048";
			}
		];
		last: [
			{
				// This looks like a bug in InDev as Page=0
				href: "/gidprojects/All?Page=0&PageSize=2048";
			}
		];
		self: [
			{
				href: "/gidprojects/All?Page=0&PageSize=2048";
			}
		];
	};
	embedded: {
		niceIndevIndevelopmentProject: Project[];
	};
}

export interface Project {
	links: {
		self: [
			{
				href: `/project/${GIDReference}`;
			}
		];
	};
	/** e.g. 01000000-0000-001B-0000-000000005402 */
	eTag: string;
	/** Project reference e.g. GID-TAG377 */
	reference: GIDReference;
	/** Acronym (usually 2 or 3 letters) for the type of the project e.g "TA", "IPG" etc */
	projectType: ProductTypeAcronym;
	/** The name of the project type e.g. "Technology appraisal guidance" for TA or "Interventional procedures guidance" for "IPG" */
	productTypeName: string;
	process: ProjectProcess | string;
	/** Project title e.g. "Atrial fibrillation - ximelagatran [ID376]" or "Clinically isolated syndrome - beta interferons and glatiramer acetate [ID109]" */
	title: string;
	status: ProjectStatus;
	projectGroup: ProjectGroup;
	/**
	 * `null` for unpublished projects or an ISO formatted date string like `2020-12-15T00:00:00`
	 */
	publishedDate: string | null;
	/**
	 * An ISO formatted date string like `2020-09-22T13:51:48.8447645`
	 */
	lastModifiedDate: string;
	/**
	 * Null for unpublished projects or an ISO formatted date string like `2020-10-05T12:27:21.5437767`
	 */
	firstGoLiveDate: string | null;
	/** The date the project was created in in dev */
	createdDate: string;
	developedAs: string | null;
	/** Almost always the same as `DevelopedAs` */
	relevantTo: string | null;
	idNumber: `${number}` | null;
	/** E.g. "BMJ Evidence Centre, BMJ Group", "ScHARR" or "Warwick Evidence" etc */
	evidenceAssessmentGroup: string | null;
	/** TODO Always seems to be an empty array */
	areasOfInterestList: [];
}

export interface InConsultationProjects {
	links: {
		self: [
			{
				href: "/inconsultationprojects?Page=0&PageSize=2048";
			}
		];
	};
	// Sometimes there are no consultations running, in which case `_embedded` is not in the feed at all
	embedded?: {
		niceIndevInconsultationProduct: Consultation[];
	};
	eTag: null;
	// Page and size is hard coded here
	page: 0;
	pageSize: 2048;
	totalItems: number;
}

export interface Consultation {
	_links: {
		self: [
			{
				href: `/guidance/${GIDReference}/consultation/html-content${
					| number
					| ""}`;
			}
		];
	};
	eTag: null;
	reference: GIDReference;
	title: string;
	consultationName: string;
	/** ISO formatted date like `2021-06-18T00:00:00` */
	startDate: string;
	/** ISO formatted date like `2021-06-18T00:00:00` */
	endDate: string;
	consultationType: string;
	resourceTitleId: string;
	projectType: ProjectProcess | string;
	productTypeName: string;
	showExpressionOfInterestSubmissionQuestion: boolean;
	developedAs: null;
	relevantTo: null;
	consultationId: number;
	/** E.g. QSD, IP etc */
	process: string;
	hasDocumentsWhichAllowConsultationComments: boolean;
	isCHTE: boolean;
	/** E.g. HSCTeam or CHTETeam */
	allowedRole: string;
	firstConvertedDocumentId: null;
	firstChapterSlugOfFirstConvertedDocument: null;
	partiallyUpdatedProjectReference: null;
	origProjectReference: null;
	areasOfInterestList: [];
}

export type Link = {
	href: string;
};

type ETag = string | null;

export type IndevFile = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	id: string;
	mimeType: string;
	fileName: string;
	length: number;
	hash: string;
	name: string;
	reference: string;
	resourceTitleId: string;
	consultationDocumentId: number;
};

export type IndevConvertedDocument = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	reference: string;
	title: string;
	resourceTitleId: string;
	projectType: string;
	productTypeName: string;
};

export type IndevResource = {
	links: {
		self: [Link];
	};
	embedded?: {
		niceIndevFile?: IndevFile;
		niceIndevConvertedDocument?: IndevConvertedDocument;
	};
	eTag: ETag;
	title: string;
	level: number;
	publishedDate: string;
	externalUrl: string | null;
	showInDocList: boolean;
	textOnly: boolean;
	consultationId: number;
	consultationDocumentId: number;
	isCurrentViewableConsultationMarkup: boolean;
	convertedDocument: boolean;
	supportsComments: boolean;
	supportsQuestions?: boolean;
};

export type IndevFileResource = IndevResource & {
	embedded: {
		niceIndevFile?: IndevFile;
		niceIndevConvertedDocument?: IndevConvertedDocument;
	};
};

export type IndevConsultation = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	reference: string;
	title: string;
	consultationName: string;
	startDate: string;
	endDate: string;
	consultationType: string;
	resourceTitleId: string;
	projectType: string;
	technologyType: string;
	productTypeName: string;
	showExpressionOfInterestSubmissionQuestion: boolean;
	developedAs: string | null;
	relevantTo: string | null;
	consultationId: number;
	process: string;
	hasDocumentsWhichAllowConsultationComments: boolean;
	isCHTE: boolean;
	allowedRole: string;
	firstConvertedDocumentId: string | null;
	firstChapterSlugOfFirstConvertedDocument: string | null;
	partiallyUpdatedProjectReference: string | null;
	origProjectReference: string | null;
	areasOfInterestList: AreaOfInterest[];
	hidden: boolean;
};

export type IndevConsultationPanel = IndevPanel & {
	embedded: {
		niceIndevConsultation: IndevConsultation;
	};
};

export type IndevPanel = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevResourceList: {
			links: {
				self: [Link];
			};
			embedded: {
				niceIndevResource: IndevResource | IndevResource[];
			};
			eTag: ETag;
			hasResources: boolean;
		};
		niceIndevConsultation?: IndevConsultation;
	};
	eTag: ETag;
	title: string;
	panelType: string;
	legacyPanel: boolean;
	originalReference: string | null;
	updateReference: string | null;
	showPanel: boolean;
};

export type IndevTimeline = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	column1: string;
	column2: string;
	additionalInfoLabel: string;
	additionalInfo: string;
	hidden: boolean;
};

export type IndevProjectTeam = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	column1: string;
	column2: string;
};

export type IndevLegacyStakeholder = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	item: string;
};

export type IndevConsultee = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	column1: string;
	column2: string;
};

export type IndevCommentator = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	column1: string;
	column2: string;
};

export type IndevEmailEnquiry = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	item: string;
};

export type IndevProcessHomepage = {
	links: {
		self: [Link];
	};
	etag: ETag;
	description: string;
	linkText: string;
} | null;

export type IndevProjectRelatedLink = {
	links: {
		self: [Link];
	};
	etag: ETag;
	column1: string;
	column2: string;
};

export type NiceIndevProjectRelatedLinkList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevProjectRelatedLink:
			| IndevProjectRelatedLink
			| IndevProjectRelatedLink[];
	};
	etag: ETag;
};

export type IndevSchedule = {
	links: {
		self: [Link];
	};
	etag: ETag;
	column1: string;
	column2: string;
	additionalInfoLabel: string;
	additionalInfo: string;
	hidden: boolean;
};

export type NiceIndevProvisionalScheduleList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevProvisionalSchedule: IndevSchedule | IndevSchedule[];
	};
	etag: ETag;
};

export type NiceIndevTimelineList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevTimeline: IndevTimeline | IndevTimeline[];
	};
};

export type NiceIndevProjectTeamList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevProjectTeam: IndevProjectTeam | IndevProjectTeam[];
	};
};

export type NiceIndevLegacyStakeholderList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevLegacyStakeholder:
			| IndevLegacyStakeholder
			| IndevLegacyStakeholder[];
	};
};

export type NiceIndevConsulteeList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevConsultee: IndevConsultee | IndevConsultee[];
	};
};

export type NiceIndevCommentatorList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevCommentator: IndevCommentator | IndevCommentator[];
	};
};

export type NiceIndevEmailEnquiryList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevEmailEnquiry: IndevEmailEnquiry | IndevEmailEnquiry[];
	};
};

export type IndevUpdate = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	updateReference: string;
};

export type NiceIndevFullUpdateList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevFullUpdate: IndevUpdate | IndevUpdate[];
	};
	eTag: ETag;
};

export type NiceIndevPartialUpdateList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevPartialUpdate: IndevUpdate | IndevUpdate[];
	};
	eTag: ETag;
};

export type IndevTopic = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	item: string;
};

export type NiceIndevTopicList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevTopic: IndevTopic | IndevTopic[];
	};
	eTag: ETag;
};

export enum ProjectType {
	"NG" = "NG",
	"CSG" = "CSG",
	"CG" = "CG",
	"DG" = "DG",
	"MPG" = "MPG",
	"ES" = "ES",
	"HST" = "HST",
	"IPG" = "IPG",
	"KTT" = "KTT",
	"LGB" = "LGB",
	"MIB" = "MIB",
	"MTG" = "MTG",
	"PH" = "PH",
	"QS" = "QS",
	"TA" = "TA",
	"PSG" = "PSG",
	"SC" = "SC",
	"SG" = "SG",
	"PIP" = "PIP",
	"APG" = "APG",
	"APA" = "APA",
	"COV" = "COV",
	"COVM" = "COVM",
	"COVT" = "COVT",
	"COA" = "COA",
	"NGC" = "NGC",
	"IND" = "IND",
	"HTE" = "HTE",
	"HUB" = "HUB",
}

export type ProjectDetail = {
	links: {
		niceIndevStakeholderRegistration: Record<string, unknown>[];
		self: [Link];
	};
	embedded: {
		niceIndevFullUpdateList?: NiceIndevFullUpdateList;
		niceIndevPartialUpdateList?: NiceIndevPartialUpdateList;
		niceIndevTopicList?: NiceIndevTopicList;
		niceIndevPanelList: {
			links: {
				self: [Link];
			};
			embedded: {
				niceIndevPanel: IndevPanel | IndevPanel[];
			};
		};
		niceIndevProvisionalScheduleList?: NiceIndevProvisionalScheduleList;
		niceIndevTimelineList?: NiceIndevTimelineList;
		niceIndevProcessHomepage?: IndevProcessHomepage;
		niceIndevProjectTeamList?: NiceIndevProjectTeamList;
		niceIndevProjectRelatedLinkList?: NiceIndevProjectRelatedLinkList;
		niceIndevLegacyStakeholderList?: NiceIndevLegacyStakeholderList;
		niceIndevConsulteeList?: NiceIndevConsulteeList;
		niceIndevCommentatorList?: NiceIndevCommentatorList;
		niceIndevEmailEnquiryList?: NiceIndevEmailEnquiryList;
	};
	eTag: ETag;
	summary: string | null;
	description: string | null;
	referralDate: string | null;
	suspendDiscontinuedReason: string | null;
	suspendDiscontinuedUrl: string | null;
	suspendDiscontinuedUrlText: string | null;
	legacyModel: boolean;
	productReference: string | null;
	evidenceAssessmentGroup: string | null;
	reference: string;
	projectType: ProjectType | null;
	technologyType: string | null;
	productTypeName: string | null;
	process: string;
	alert: string | null;
	title: string;
	content: string | null;
	status: ProjectStatus;
	projectGroup: ProjectGroup;
	publishedDate: string;
	lastModifiedDate: string;
	firstGoLiveDate: string;
	createdDate: string;
	developedAs: string;
	relevantTo: string | null;
	idNumber: string | null;
	areasOfInterestList: [];
	topicSelectionDecision: string;
	topicSelectionReason: string | null;
	topicSelectionDecisionDate: string | null;
	topicSelectionFurtherInfo: string | null;
	indicatorSubTypes: [];
};

export type resourceInPageNavLink = {
	slug: string;
	title: string;
};

export type niceIndevConvertedDocumentChapter = {
	slug: string;
	title: string;
	href: string;
};

export type niceIndevConvertedDocumentSection = {
	slug: string;
	title: string;
};

export type niceIndevConvertedDocument = {
	_links?: {
		self: [
			{
				href: string;
			}
		];
	};
	eTag?: string | null;
	content: string;
	pdfLink: string | null;
	title?: string;
	chapterTitle?: string;
	chapters: niceIndevConvertedDocumentChapter[];
	sections?: niceIndevConvertedDocumentSection[];
};
