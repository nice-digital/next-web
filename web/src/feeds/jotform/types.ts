export type FormID = `${number}`;

export type Duration = `${number}ms`;

export type NotFoundResponse = {
	responseCode: 404;
	message: string;
	content: "";
	duration: Duration;
	info: string;
};

export type NotAuthorizedResponse = {
	responseCode: 401;
	message: string;
	content: "";
	duration: Duration;
	info: string;
};

export type FormStatus = "ENABLED" | "DISABLED";

export type FormType = "LEGACY" | "CARD";

export type FormSuccessResponse = {
	responseCode: 200;
	message: "success";
	content: {
		id: string;
		username: string;
		title: string;
		height: number;
		status: FormStatus;
		/** Format like "2023-03-21 09:27:18" */
		created_at: string;
		updated_at: string;
		last_submission: string;
		new: `${number}`;
		count: `${number}`;
		type: FormType;
		favorite: `${number}`;
		archived: `${number}`;
		url: string;
	};
	duration: Duration;
};

export type GetFormResponse =
	| NotAuthorizedResponse
	| NotFoundResponse
	| FormSuccessResponse;
