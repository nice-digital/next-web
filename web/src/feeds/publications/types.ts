import type { Except, ReadonlyDeep } from "type-fest";

export enum FeedPath {
	ProductsLite = "/feeds/products-lite",
	ProductTypes = "/feeds/producttypes",
	AreasOfInterest = "/feeds/areaofinteresttypes",
	ProductDetail = "/feeds/product/",
}

export enum ProductStatus {
	Published = "Published",
	Withdrawn = "Withdrawn",
}

export enum ProductGroup {
	Guideline = "Guideline",
	Guidance = "Guidance",
	Advice = "Advice",
	Standard = "Standard",
	Corporate = "Corporate",
	Other = "Other",
}

export const enum ProductTypeAcronym {
	/** Antimicrobial prescribing evidence summaries */
	"APA" = "APA",
	/** Antimicrobial prescribing guidelines */
	"APG" = "APG",
	/** Clinical guidelines */
	"CG" = "CG",
	/** COVID-19 rapid evidence summaries */
	"COA" = "COA",
	/** COVID-19 rapid guidelines */
	"COV" = "COV",
	/** COVID-19 rapid MIBs */
	"COVM" = "COVM",
	/** COVID-19 tests */
	"COVT" = "COVT",
	/** Cancer service guidelines */
	"CSG" = "CSG",
	/** Diagnostics guidance */
	"DG" = "DG",
	/** Corporate documents (corporate)  */
	"ECD" = "ECD",
	/** Evidence summaries (advice) */
	"ES" = "ES",
	/** Highly specialised technologies guidance */
	"HST" = "HST",
	/** Indicators */
	"IND" = "IND",
	/** Interventional procedures guidance */
	"IPG" = "IPG",
	/** Key therapeutic topics (advice) */
	"KTT" = "KTT",
	/** Local government briefing (advice) */
	"LGB" = "LGB",
	/** Medtech innovation briefings (advice) */
	"MIB" = "MIB",
	/** Medicines practice guidelines */
	"MPG" = "MPG",
	/** Medical technologies guidance */
	"MTG" = "MTG",
	/** NICE guidelines */
	"NG" = "NG",
	/** Public health guidelines */
	"PH" = "PH",
	/** Public involvement programmes (advice) */
	"PIP" = "PIP",
	/** Process and methods guides (corporate) */
	"PMG" = "PMG",
	/** Patient safety guidance */
	"PSG" = "PSG",
	/** Quality standards (standards) */
	"QS" = "QS",
	/** Social care guidelines */
	"SC" = "SC",
	/** Safe staffing guidelines */
	"SG" = "SG",
	/** Technology appraisal guidance */
	"TA" = "TA",
}

export enum AreaOfInterestAcronym {
	/**
	 * Covid-19
	 */
	COV = "COV",
	/**
	 * Antimicrobial prescribing
	 */
	AMP = "AMP",
}

/** Some product types can act as parents for sub-types */
export type ParentProductTypeAcronym =
	// ES (Evidence summaries) have sub types like APA (Antimicrobial prescribing evidence summaries) and COA (COVID-19 rapid evidence summaries)
	| ProductTypeAcronym.ES
	// MIBs (Medtech innovation briefings) have sub types like COA (COVID-19 rapid MIBs) and COVT (COVID-19 tests)
	| ProductTypeAcronym.MIB;

/** Only certain types can be used for 'developed as' */
export type DevelopedAsProductTypeAcronym =
	| ProductTypeAcronym.APA
	| ProductTypeAcronym.APG
	| ProductTypeAcronym.CG
	| ProductTypeAcronym.COA
	| ProductTypeAcronym.COV
	| ProductTypeAcronym.COVM
	| ProductTypeAcronym.MPG
	| ProductTypeAcronym.PH
	| ProductTypeAcronym.SC
	| ProductTypeAcronym.SG;

/** Only certain types can be used for 'relevant to' */
export type RelevantToProductTypeAcronym =
	| ProductTypeAcronym.APA
	| ProductTypeAcronym.APG
	| ProductTypeAcronym.COV
	| ProductTypeAcronym.CSG
	| ProductTypeAcronym.MPG
	| ProductTypeAcronym.PH
	| ProductTypeAcronym.SC
	| ProductTypeAcronym.SG;

export type Link = {
	readonly href: string;
};

interface EmptyLinks {
	/** Empty object that's never used but in here for completeness */
	readonly self: [Record<string, never>];
}

export type BaseFeedItem = ReadonlyDeep<{
	_links: EmptyLinks;
	/** ETag is always null so kind of pointless but kept here for completeness */
	eTag: null;
	/** Full ISO date string like `2021-04-15T08:18:13.7945978Z` */
	lastModified: string;
}>;

/**
 * The raw object that comes back from the feed
 */
export type ProductLiteRaw = BaseFeedItem &
	ReadonlyDeep<{
		_links: EmptyLinks & {
			"nice.publications:productfeed": Link[];
		};
		ETag: null;
		Id: string;
		Title: string;
		ProductStatus: ProductStatus;
		ProductType: ProductTypeAcronym;
		/** ISO date string like `2017-07-06T00:00:00`. Notice it's a rounded time because it's set manually by an editor. */
		PublishedDate: string;
		/** ISO date string like `2017-07-06T00:00:00` or `2021-05-21T10:54:52.4655487`. */
		LastMajorModificationDate: string;
		AreasOfInterestList: AreaOfInterestAcronym[];
		DevelopedAs: DevelopedAsProductTypeAcronym | null;
		RelevantTo: RelevantToProductTypeAcronym[];
		ProductGroup: ProductGroup;
	}>;

/** A product lite from the feed, but with redundant properties removed */
export type ProductLite = Except<ProductLiteRaw, "ETag" | "_links">;

export type ProductType = BaseFeedItem &
	ReadonlyDeep<{
		enabled: boolean;
		name: string;
		pluralName: string;
		identifierPrefix: ProductTypeAcronym;
		group: ProductGroup;
		parent: ParentProductTypeAcronym | "";
	}>;

export type AreaOfInterest = BaseFeedItem &
	ReadonlyDeep<{
		Enabled: boolean;
		Name: string;
		PluralName: string;
	}>;

type ETag = string | null;

// axios-case-converter maps key property names like nice.publications:area-of-interest-type-list to nicePublicationsAreaOfInterestTypeList
type EmbeddedKey = `nicePublications${string}`;

type Embedded<TKey extends EmbeddedKey, TInner> = {
	embedded: { [key in TKey]: TInner };
};

type FeedContentInner<TEmbeddedInner extends EmbeddedKey, TItemType> = {
	_links: {
		self: [Link];
	};
	eTag: ETag;
} & Embedded<TEmbeddedInner, TItemType[]>;

type FeedContent<
	TEmbeddedOuter extends EmbeddedKey,
	TEmbeddedInner extends EmbeddedKey,
	TItemType
> = {
	_links: {
		self: [Link];
	};
	eTag: ETag;
	lastModified: string;
} & Embedded<TEmbeddedOuter, FeedContentInner<TEmbeddedInner, TItemType>>;

export type AreasOfInterestList = FeedContent<
	"nicePublicationsAreaOfInterestTypeList",
	"nicePublicationsAreaOfInterestType",
	AreaOfInterest
>;

export type ProductTypeList = FeedContent<
	"nicePublicationsProductTypeList",
	"nicePublicationsProductType",
	ProductType
>;

export type ProductListLite = FeedContent<
	"nicePublicationsProductListLite",
	"nicePublicationsProductLite",
	ProductLiteRaw
>;

export type ProductChapter = {
	title: string;
	url: string;
};

export type HTMLChapterContentInfo = {
	title: string;
	chapterSlug: string;
	_links: { self: [Link] };
};

export type HTMLChapterContent = {
	eTag: string;
	content: string;
	title: string;
	reference: string;
	partId: number;
	chapterSlug: string;
};

export type HTMLContent = Embedded<
	"nicePublicationsHtmlChapterContentInfo",
	HTMLChapterContentInfo[]
>;

// export type HTMLContent = {
// 	_embedded: {
// 		"nice.publications:html-chapter-content-info": HTMLChapterContentInfo[];
// 	};
// };

export type FileContent<TFileExtension extends "pdf" | "mobi" | "epub"> = {
	fileName: `${string}.${TFileExtension}`;
	length: number;
	_links: { self: [Link] };
	eTag: string | null;
	uid: number;
	id: string;
	mimeType: string;
	hash: string;
	name: string;
};

export type UploadAndConvertContentPart = {
	embedded: {
		nicePublicationsHtmlContent: HTMLContent;
		nicePublicationsPdfFile: FileContent<"pdf">;
		nicePublicationsMobiFile: FileContent<"mobi">;
		nicePublicationsEpubFile: FileContent<"epub">;
	};
};

// export type ContentPartListOLD = {
// 	"nice.publications:content-part-list": {
// 		embedded: UploadAndConvertContentPart;
// 	};
// };

export type ContentPartList = Embedded<
	"nicePublicationsUploadAndConvertContentPart",
	UploadAndConvertContentPart
>;

export type ProductDetail = {
	chapterHeadings?: ProductChapter[];
	// embedded: ContentPartList;
	id: string;
	lastUpdatedDate?: string;
	productType: string;
	publishedDate?: string;
	summary?: string;
	title: string;
} & Embedded<"nicePublicationsContentPartList", ContentPartList>;

export type ErrorResponse = {
	StatusCode: string;
	Message: string;
};
