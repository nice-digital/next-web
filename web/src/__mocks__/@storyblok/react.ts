export const storyblokInit = (): null => null;

export const apiPlugin = (): null => null;
