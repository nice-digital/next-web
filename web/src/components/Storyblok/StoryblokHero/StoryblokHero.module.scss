@use '@nice-digital/nds-core/scss/media-queries';

.heroBreadcrumbs {
  @include media-queries.mq($from: md) {
    width: 66%;
  }
}

.hero {
  margin-bottom: 0;
}

