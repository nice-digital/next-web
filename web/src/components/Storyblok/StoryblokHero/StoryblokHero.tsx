import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";
import { Hero } from "@nice-digital/nds-hero";

import { StoryblokButtonLink } from "@/components/Storyblok/StoryblokButtonLink/StoryblokButtonLink";
import { type Breadcrumb as TypeBreadcrumb } from "@/types/Breadcrumb";
import { type HeroStoryblok } from "@/types/storyblok";

export interface HeroBlokProps {
	blok: HeroStoryblok;
	breadcrumbs?: TypeBreadcrumb[];
}

export const StoryblokHero = ({
	blok,
	breadcrumbs,
}: HeroBlokProps): React.ReactElement => {
	const { image, cta } = blok;

	const BreadcrumbComponent = breadcrumbs?.length ? (
		<Breadcrumbs>
			{[{ title: "Home", path: "/" }, ...breadcrumbs].map((breadcrumb) => (
				<Breadcrumb key={breadcrumb.title} to={breadcrumb.path}>
					{breadcrumb.title}
				</Breadcrumb>
			))}
		</Breadcrumbs>
	) : undefined;

	const imageUrl = `${image?.filename}/m/` || undefined;

	const action = cta?.length ? (
		<StoryblokButtonLink button={cta[0]} />
	) : undefined;

	return (
		<Hero
			title={blok.title}
			intro={blok.summary || undefined}
			header={BreadcrumbComponent}
			image={imageUrl}
			actions={action}
		/>
	);
};
