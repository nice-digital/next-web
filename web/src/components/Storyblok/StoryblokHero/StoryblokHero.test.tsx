import { screen } from "@testing-library/react";

import { render } from "@/test-utils/rendering";
import { mockImageAsset } from "@/test-utils/storyblok-data";
import { type Breadcrumb as TypeBreadcrumb } from "@/types/Breadcrumb";

import { StoryblokHero, type HeroBlokProps } from "./StoryblokHero";

export const heroProps: HeroBlokProps = {
	blok: {
		title: "Holding out for a hero",
		image: mockImageAsset,
		_uid: "123245789",
		component: "hero",
		cta: [
			{
				_uid: "123456789",
				component: "buttonLink",
				text: "Click me",
				link: {
					_uid: "123456789",
					url: "https://example.com",
				},
				variant: "primary",
			},
		],
		summary: "The hero we need",
		description: "The hero we deserve",
	},
};

const breadcrumbs: TypeBreadcrumb[] = [
	{
		title: "Home",
		path: "/",
	},
	{
		title: "All about bacon",
		path: "/bacon",
	},
	{
		title: "Wonderful bacon additive",
		path: "/bacon/maple-syrup",
	},
];

export const heroBreadcrumbProps = {
	...heroProps,
	breadcrumbs,
};

describe("Storyblok Hero", () => {
	it("should match snapshot without breadcrumbs", () => {
		const { container } = render(<StoryblokHero {...heroProps} />);
		expect(container).toMatchSnapshot();
	});

	it("should match snapshot with breadcrumbs", () => {
		const { container } = render(<StoryblokHero {...heroBreadcrumbProps} />);
		expect(container).toMatchSnapshot();
	});

	it("should render breadcrumbs if supplied", () => {
		render(<StoryblokHero {...heroBreadcrumbProps} />);
		expect(screen.getAllByRole("link").length).toBe(breadcrumbs.length);
	});

	it("should apply the 'hero--dark' class when theme is 'impact'", () => {
		const impactThemedProps: HeroBlokProps = {
			...heroProps,
			blok: { ...heroProps.blok, theme: "impact" },
		};
		render(<StoryblokHero {...impactThemedProps} />);
		expect(screen.getByTestId("hero-container")).toHaveClass("hero--dark");
	});

	it("should not apply the 'hero--dark' class when theme is 'subtle'", () => {
		const subtleThemedProps: HeroBlokProps = {
			...heroProps,
			blok: { ...heroProps.blok, theme: "subtle" },
		};

		render(<StoryblokHero {...subtleThemedProps} />);
		expect(screen.getByTestId("hero-container")).not.toHaveClass("hero--dark");
	});

	it("should not apply the 'hero--dark' class when theme is undefined", () => {
		render(<StoryblokHero {...heroProps} />);
		expect(screen.getByTestId("hero-container")).not.toHaveClass("hero--dark");
	});
});
