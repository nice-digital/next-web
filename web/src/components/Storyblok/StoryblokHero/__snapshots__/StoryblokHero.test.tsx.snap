// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyblok Hero should match snapshot with breadcrumbs 1`] = `
<div>
  <div
    class="undefined undefined"
  >
    <div />
    <main>
      <div
        class="container"
        data-component="container"
      >
        <div
          class="hero"
          data-component="hero"
        >
          <div
            class="hero__container"
          >
            <nav
              aria-label="Breadcrumbs"
              data-component="breadcrumbs"
              role="navigation"
            >
              <ol
                class="breadcrumbs"
              >
                <li
                  class="breadcrumbs__crumb"
                >
                  <a
                    href="/"
                  >
                    Home
                  </a>
                </li>
                <li
                  class="breadcrumbs__crumb"
                >
                  <a
                    href="/bacon"
                  >
                    All about bacon
                  </a>
                </li>
                <li
                  class="breadcrumbs__crumb"
                >
                  <a
                    href="/bacon/maple-syrup"
                  >
                    Wonderful bacon additive
                  </a>
                </li>
              </ol>
            </nav>
            <script
              type="application/ld+json"
            >
              {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Home"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/bacon",
        "name": "All about bacon"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "/bacon/maple-syrup",
        "name": "Wonderful bacon additive"
      }
    }
  ]
}
            </script>
            <div
              class="hero__body"
            >
              <h1
                class="hero__title"
              >
                Holding out for a hero
              </h1>
              <p
                class="hero__intro"
              >
                The hero we need
              </p>
              <p
                class="hero__description"
              >
                The hero we deserve
              </p>
              <div
                class="hero__actions"
              />
            </div>
          </div>
          <div
            class="hero__image-container"
            style="background-image: url(http://local-host-test-nice-org.com/f/243782/420x234/7a576133e7/register-for-board-meeting.PNG/m/867x0/filters:format%28jpeg%29:quality%2880%29/m/filters:quality%2880%29);"
          />
        </div>
      </div>
    </main>
    <noscript />
  </div>
</div>
`;

exports[`Storyblok Hero should match snapshot without breadcrumbs 1`] = `
<div>
  <div
    class="undefined undefined"
  >
    <div />
    <main>
      <div
        class="container"
        data-component="container"
      >
        <div
          class="hero"
          data-component="hero"
        >
          <div
            class="hero__container"
          >
            <div
              class="hero__body"
            >
              <h1
                class="hero__title"
              >
                Holding out for a hero
              </h1>
              <p
                class="hero__intro"
              >
                The hero we need
              </p>
              <p
                class="hero__description"
              >
                The hero we deserve
              </p>
              <div
                class="hero__actions"
              />
            </div>
          </div>
          <div
            class="hero__image-container"
            style="background-image: url(http://local-host-test-nice-org.com/f/243782/420x234/7a576133e7/register-for-board-meeting.PNG/m/867x0/filters:format%28jpeg%29:quality%2880%29/m/filters:quality%2880%29);"
          />
        </div>
      </div>
    </main>
    <noscript />
  </div>
</div>
`;
